ballerina/variable_validator.bal

```diff
@@ -154,4 +172,47 @@ class VariableValidator {
             self.errors.push(getErrorDetailRecord(message, location)); 
         }
     }
+
+    isolated function checkVariableUsageCompatibility(__InputValue[] input, parser:VariableDefinition varDef,
+                                                      parser:ArgumentNode argNode) {
+        __InputValue? inputValue = getInputValueFromArray(input, argNode.getName());
+        __Type? actualType = self.getTypeRecordFromTypeName(varDef.kind);
+        if inputValue is __InputValue && actualType is __Type {
+            if getOfType(inputValue.'type)?.name != getOfType(actualType)?.name {
+                string message = string`Variable "${<string>argNode.getVariableName()}" of type "${varDef.kind}" `+
+                string`used in position expecting type "${getTypeNameFromType(inputValue.'type)}".`;
+                self.errors.push(getErrorDetailRecord(message, argNode.getLocation()));
+            } else if inputValue.'type.kind == NON_NULL && actualType.kind != NON_NULL && varDef?.defaultValue is () {
+                string message = string`Variable "${<string>argNode.getVariableName()}" of type "${varDef.kind}" `+
+                string`used in position expecting type "${getTypeNameFromType(inputValue.'type)}".`;
+                self.errors.push(getErrorDetailRecord(message, argNode.getLocation()));
+            }
+        }
+    }
+
+    isolated function getTypeRecordFromTypeName(string typeName) returns __Type? {
+        //TODO: Improve for 'List' type variables
+        if typeName.endsWith("!") {
```

Comment:
This is checked twice. Once here, and once again in `getOfTypeNameFromTypeName`. We can simplify this.

Reply:
What's the resolution for this?

Using the same string operations on the same string twice does not seems good. We can use this function to get the type, and if we want the type name we can use `getOfType()?.name` which is efficient.

Reply:
Then we may have to use string casting there since `getOfType()?.name` is optional field and bring checks in `checkVariableUsageCompatibility` into `visitArgument` since we don't have exact input type in `visitArgument`.